src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -60,16 +90,71 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
         public IEnumerable<ConfigurationClient> GetAvailableClients(DateTimeOffset time)
```

Comment:
I think this needs to be IAsyncEnumerable and contain replica discovery within.

Reply:
I have a different idea on this, SRV lookup comes with costs, we only need to perform SRV lookup when we have to.
The only situation of having to query SRV happens when all existing clients are not available, as long as any client in the `_clients` list is available, we don't need to bother to query the SRV.

There are two different cases of what we may get from `GetAvailableClients()`.
1. We never performed SRV lookup, which means we can just get clients that passed by the user. If any user's client is available , we don't need to query SRV.
2. We have performed SRV lookup in the previous refresh, means the discovered clients should already added to the `_clients` list, `GetAvailableClients()` will return all the available user's clients and discovered clients. user's clients are prior to the discovered clients in this list. So user's list would always be picked up first.

Reply:
If this method returns an empty client list, means all existing clients fall into backoff. I think we still don't need to query SRV in this case, because we know we already have performed SRV lookup before all clients fall into backoff, it's not reasonable to query SRV during the all clients backoff. Once any  client wake up and it still get failoverable failure, we can consider perform a new round SRV

Reply:
We don't need to get fallback clients every time and we can still take my suggestion.

Reply:
Something like below

```
public async IAsyncEnumerable GetAvailableClients()
{
    now = DateTimeOffset.UtcNow

    foreach (clientWrapper in _clients)
    {
        if (clientWrapper.BackOffEndTime <= DateTimeOffset.UtcNow)
        {
            yield return clientWrapper.Client;
        }
    }

    if (now >= lastFallbackClientRefresh + FallbackClientRefreshInterval)
    {
        await RefreshFallBackClients();
    }

    foreach (clientWrapper in _fallbackClients)
    {
        if (clientWrapper.BackOffEndTime <= DateTimeOffset.UtcNow)
        {
            yield return clientWrapper.Client;
        }
    }
}
```

Reply:
You are saying that we can apply backoff logic to fallback clients refreshing as well. 

The way I currently implement is equivalent to applying the existing exponential backoff algorithm to fallback clients lookup.

I'm good with your suggestion about internalizing all knowledge about obtaining clients in the manager,  the obtaining clients logic does not invade into `AzureAppConfigurationProvider`

How about changing to this:
1. On the app's start-up, perform once fallback clients lookup
2. During the app running, only perform lookup when:
      a.Get no available clients from `_clients` list (all existing clients are backoff)
      b. now >= _lastFallbackClientRefresh + FallbackClientRefreshInterval

I'd like to give a constant value (5 min) to `FallbackClientRefreshInterval`, because the maximum value of client backoff is 10 mins, just take the median of it,  are you good with it?

Reply:
> How about changing to this:
On the app's start-up, perform once fallback clients lookup
During the app running, only perform lookup when:
a.Get no available clients from _clients list (all existing clients are backoff)
b. now >= _lastFallbackClientRefresh + FallbackClientRefreshInterval

Sounds good, but I'd add one more condition.

If all clients (static and dynamic) are backed off, perform a lookup to see if any new ones have been added, even if a refresh is not due.

Reply:
I just pushed a new commit, please take a look. Refresh fallback clients when:
 1. It's the initial attempt to query available clients on app start-up.
 2. At least one client is available in the last attempt, it's the first attempt that all static and dynamic clients are backed off. (Will perform lookup immediately even if refresh is not due)
 3. It's not the first attempt that all clients have backed off, as well as it's due to refresh fallback clients. (ensure lookup won't be too frequent when all existing clients keep backoff )